### PR TITLE
api: use <scheme,provider> map in nameResoverRegistry

### PR DESCRIPTION
**Pull request by YifeiZhuang**
Part of https://github.com/grpc/grpc-java/issues/7133.
It makes scheme matching more explicit in name resolver API. It slightly changed the existing behaviour because in this PR we pick effective providers early/when refreshing.
But it should not change effectiveness of e.g. `SecretGrpclbNameResolverProvider`.

### CODE REVIEW

File: xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java [hunk 1]
```diff
@@ -88,8 +88,11 @@
   private TlsContextManagerImpl tlsContextManagerForServer;
 
   @Before
-  public void setUp() throws IOException {
+  public void setUp() throws Exception {
     port = XdsServerTestHelper.findFreePort();
+    URI expectedUri = new URI("sdstest://localhost:" + port);
+    fakeNameResolverFactory = new FakeNameResolverFactory.Builder(expectedUri).build();
+    NameResolverRegistry.getDefaultRegistry().register(fakeNameResolverFactory);
```
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java:95]: Is this change required because of the change to the Registry or is this just an improvement to this test?

Comment (YifeiZhuang) [reply xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java:95]: oh this is to fix a bug that the test case registered `fakeNameResolverFactory` twice. Moving `register` to `@Before` guarantees it registered only once. Otherwise the leaked provider will cause the next test fail.

File: api/src/main/java/io/grpc/NameResolverRegistry.java [hunk 2]
```diff
@@ -149,23 +151,17 @@ public static synchronized NameResolverRegistry getDefaultRegistry() {
...
 
     @Override
     public String getDefaultScheme() {
-      List<NameResolverProvider> providers = providers();
+      LinkedHashMap<String, NameResolverProvider> providers = providers();
       if (providers.isEmpty()) {
         return "unknown";
       }
-      return providers.get(0).getDefaultScheme();
+      return providers.entrySet().iterator().next().getKey();
```
Comment (ejona86) [inline api/src/main/java/io/grpc/NameResolverRegistry.java:164]: I don't think this does what you are hoping. This does not get the first entry with the highest priority, as items are inserted into the map in `allProviders`-order.

It seems easiest to just compute this value within `refreshProviders()`, either manually within the loop or using something like `Collections.max()` (although max() doesn't guarantee any particular behavior when multiple entries are maximal).

File: api/src/main/java/io/grpc/NameResolverRegistry.java [hunk 3]
```diff
@@ -81,16 +82,16 @@ public synchronized void deregister(NameResolverProvider provider) {
...
-    // Sort descending based on priority.
-    // sort() must be stable, as we prefer first-registered providers
-    Collections.sort(providers, Collections.reverseOrder(new Comparator<NameResolverProvider>() {
-      @Override
-      public int compare(NameResolverProvider o1, NameResolverProvider o2) {
-        return o1.priority() - o2.priority();
+    LinkedHashMap<String, NameResolverProvider> refreshedProviders = new LinkedHashMap<>();
+    // We prefer first-registered providers
+    for (NameResolverProvider provider : allProviders) {
+      String scheme = provider.getDefaultScheme();
```
Comment (ejona86) [inline api/src/main/java/io/grpc/NameResolverRegistry.java:88]: This method doesn't have the correct semantics; I think we need a new API for this, like `getScheme()`. But we can make the default implementation of `getScheme()` delegate to `getDefaultScheme()`.

File: api/src/main/java/io/grpc/NameResolverRegistry.java [hunk 4]
```diff
@@ -47,9 +47,10 @@
 
   @GuardedBy("this")
   private final LinkedHashSet<NameResolverProvider> allProviders = new LinkedHashSet<>();
-  /** Immutable, sorted version of {@code allProviders}. Is replaced instead of mutating. */
```
Comment (ejona86) [inline api/src/main/java/io/grpc/NameResolverRegistry.java:50]: Why delete the "is replaced instead of mutating" comment? That seems all the more important now that it is a LinkedHashMap and so can't easily be made unmodifiable.

File: xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java (hunk 1 above)
Comment (YifeiZhuang) [reply xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java:95]: oh do you mean [this](https://github.com/grpc/grpc-java/pull/8323/files#diff-cb9cb5fa575a01d40895ffc9883bc1026b65a40fb426f311f26f47f01a3e63b6L98)?, it's already like that. The previous error comes from picking the provider that didn't deregister previously thus failing the test code verification, it has nothing to do with the registry component.

File: api/src/main/java/io/grpc/NameResolverRegistry.java (hunk 2 above)
Comment (YifeiZhuang) [reply api/src/main/java/io/grpc/NameResolverRegistry.java:164]: yea we probably should return the highest priority provider scheme as the default scheme. Then I guess we do not need linkedhashmap anymore just a normal immutable hashmap. `allProviders` will guarantee we get the consistent default scheme.

File: xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java (hunk 1 above)
Comment (ejona86) [reply xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java:95]: The link doesn't take me anywhere useful. But I do see there is a deregister within the `@After`. So that means getStub() is called twice. Seems that is only done in `mtlsClientServer_changeServerContext_expectException()`. Okay, this seems fine.

**Review by ejona86 — APPROVED** [review]
I've audited the internal NameResolverProviders and all have a `SCHEME.equals(uri.getScheme())` check, so I don't think this will break anything internally.
